File: python_wrapper/env_wrapper.py
import argparse
import multiprocessing
import os
import random
from typing import Any, Dict, List, Optional, Tuple


class EnvWrapper:
    MAX_READ_SIZE = 4096
    END_TOKEN = "OK"

    def __init__(self, pipe_name, launch_command) -> None:
        self.in_pipe_name = pipe_name + "_out"
        self.out_pipe_name = pipe_name + "_in"

        self.launch_command = launch_command
        self.env_process = None
        self.in_pipe = None
        self.out_pipe = None

    def _cleanEnv(self):
        if self.env_process is not None:
            self.env_process.terminate()

    def _makePipe(self):
        os.mkfifo(self.in_pipe_name)
        os.mkfifo(self.out_pipe_name)
        pipe_flag = os.O_SYNC | os.O_CREAT | os.O_RDWR
        self.in_pipe = os.open(self.in_pipe_name, pipe_flag)
        self.out_pipe = os.open(self.out_pipe_name, pipe_flag)

    def _removePipe(self):
        if self.in_pipe is not None \
                and self.out_pipe is not None:
            os.close(self.in_pipe)
            os.close(self.out_pipe)
            os.remove(self.in_pipe_name)
            os.remove(self.out_pipe_name)

    @staticmethod
    def _parseObs(obs) -> Optional[Dict]:
        obs = obs.decode(encoding="ASCII")
        if obs == EnvWrapper.END_TOKEN:
            return None

        def _parseBaseInfo(lines: List[str]):
            frame_id, money = tuple(map(int, lines.pop(0).split()))
            return {
                "frame_id": frame_id,
                "money": money,
            }

        def _parseStation(lines: List[str]):
            station_specs = {
                "station_type": int,
                "loc_x": float,
                "loc_y": float,
                "remain_time": int,
                "input_status": int,
                "output_status": int,
            }

            num_stations = int(lines.pop(0))
            stations = []
            for _ in range(num_stations):
                items = lines.pop(0).split()
                station = {
                    key: station_specs[key](value)
                    for key, value in zip(station_specs.keys(), items)
                }
                stations.append(station)
            return {"stations": stations}

        def _parseRobots(lines: List[str]):
            robot_specs = {
                "station_id": int,
                "item_type": int,
                "time_coef": float,
                "momentum_coef": float,
                "angular_speed": float,
                "line_speed_x": float,
                "line_speed_y": float,
                "theta": float,
                "loc_x": float,
                "loc_y": float,
            }
            robots = []
            for _ in range(4):
                items = lines.pop(0).split()
                robot = {
                    key: robot_specs[key](value)
                    for key, value in zip(robot_specs.keys(), items)
                }
                robots.append(robot)
            return {"robots": robots}

        obs_dict: Dict[str, Any] = {}
        lines = obs.split('\n')
        obs_dict.update(_parseBaseInfo(lines))
        obs_dict.update(_parseStation(lines))
        obs_dict.update(_parseRobots(lines))
        return obs_dict

    def reset(self):
        self._cleanEnv()
        self._removePipe()
        self._makePipe()

        self.env_process = multiprocessing.Process(
            target=lambda cmd: os.system(cmd), args=(self.launch_command, ))
        print(f"[INFO]: Launch Environment: {self.launch_command}")
        self.env_process.start()

        assert self.in_pipe is not None
        observation = os.read(self.in_pipe, self.MAX_READ_SIZE * 10)
        return {"map": observation.decode(encoding="ASCII")}

    def recv(self) -> Tuple[Optional[Dict], bool]:
        """
        Returns:
            observation (object): agent's observation of the current environment
            done (boolean): whether the episode has ended, in which case further step() calls will return None
        """
        assert self.in_pipe is not None
        observation = os.read(self.in_pipe, self.MAX_READ_SIZE)
        obs = self._parseObs(observation)
        return obs, obs is None

    def send(self, actions: List[str]) -> None:
        assert self.out_pipe is not None
        os.write(self.out_pipe, '\n'.join(actions).encode(encoding="ASCII"))

    def close(self):
        self._cleanEnv()
        self._removePipe()


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--map-id", default="maps/1.txt")
    parser.add_argument("--env-binary-name", default="./Robot_fast")
    parser.add_argument("--env-args", default="-d")
    parser.add_argument("--env-wrapper-name", default="./env_wrapper")
    parser.add_argument("--pipe-name", default=f"/tmp/pipe_{random.random()}")
    args = parser.parse_args()
    print(args)

    from demo.simple_agent import SimpleAgent
    launch_command = f"{args.env_binary_name} {args.env_args} "\
        f"-m {args.map_id} \"{args.env_wrapper_name} {args.pipe_name}\""
    env = EnvWrapper(args.pipe_name, launch_command)
    obs = env.reset()
    agent = SimpleAgent(obs)
    for _ in range(100):
        obs, done = env.recv()
        assert not done
        action = agent.act(obs)
        env.send(action)

    for _ in range(2):
        obs = env.reset()
        while True:
            obs, done = env.recv()
            if done:
                break
            action = agent.act(obs)
            env.send(action)

    print("[INFO]: Waiting for environment to close...")
    env.close()

    print("[INFO]: Done!")

File: python_wrapper/env_wrapper.cpp
#include <iostream>
#include <cstring>
#include <string>
#include <sstream>

#include <fcntl.h>
#include <unistd.h>

static const unsigned MAX_RECV_SIZE = 1024; // bytes
static const bool VERBOSE = false;
static const std::string END_TOKEN = "OK";

class NamedPipe
{
private:
    std::string pipe_name;
    int pipe_fd;
    char buf[MAX_RECV_SIZE];

public:
    NamedPipe(const std::string &name) : pipe_name(name)
    {
        // link to the pipe
        pipe_fd = open(pipe_name.c_str(), O_RDWR);
        if (pipe_fd == -1)
        {
            std::cerr << "Failed to open pipe: " << pipe_name << std::endl;
            exit(EXIT_FAILURE);
        }
    }

    ~NamedPipe()
    {
        ::close(pipe_fd);
    }

    void write(const std::string &msg)
    {
        if (::write(pipe_fd, msg.c_str(), msg.size()) == -1)
        {
            std::cerr << "Failed to write to pipe: " << pipe_name << std::endl;
            exit(EXIT_FAILURE);
        }
    }

    std::string read()
    {
        memset(buf, 0, sizeof(buf));
        if (::read(pipe_fd, buf, sizeof(buf)) == -1)
        {
            std::cerr << "Failed to read from pipe: " << pipe_name << std::endl;
            exit(EXIT_FAILURE);
        }
        return std::string(buf);
    }
};

class Agent
{
private:
    long long frame_id;
    std::string observation;
    NamedPipe &in_pipe, &out_pipe;

public:
    Agent(NamedPipe &in_pipe, NamedPipe &out_pipe)
        : in_pipe(in_pipe), out_pipe(out_pipe)
    {
        // read the map
        observation.clear();
        std::string line;
        while (std::getline(std::cin, line))
        {
            if (line == END_TOKEN)
                break;
            observation += line + "\n";
        }

        // forward to pipe
        out_pipe.write(observation);

        if (VERBOSE)
            std::cerr << "Map received: \n"
                      << observation << std::endl;

        std::cout << END_TOKEN << std::endl;
    }

    void recv()
    {
        // read the observation
        observation.clear();
        std::string line;
        while (std::getline(std::cin, line))
        {
            if (line == END_TOKEN)
                break;
            observation += line + "\n";
        }

        // forward to pipe
        out_pipe.write(observation);

        std::stringstream ss(observation);
        ss >> frame_id; // read the frame id

        if (VERBOSE)
            std::cerr << "Observation received: \n"
                      << observation << std::endl;
    }

    void send()
    {
        // read the action from pipe
        std::string action = in_pipe.read();

        if (VERBOSE)
            std::cerr << "Action received: \n"
                      << action << std::endl;

        std::cout << frame_id << "\n"
                  << action << '\n'
                  << END_TOKEN << std::endl;
    }
};

int main(const int argc, const char *argv[])
{
    if (argc < 2)
    {
        std::cerr << "Usage: " << argv[0] << " <pipename>" << std::endl;
        return 1;
    }
    std::string pipe_name(argv[1]);
    std::cerr << "Input pipe: " << pipe_name << "_in" << '\n'
              << "Output pipe: " << pipe_name << "_out" << std::endl;

    NamedPipe in_pipe(pipe_name + "_in");
    NamedPipe out_pipe(pipe_name + "_out");

    Agent agent(in_pipe, out_pipe);

    int loop_count = 0;
    while (!std::cin.eof())
    {
        if (VERBOSE)
        {
            std::cerr << "Loop count: " << loop_count << std::endl;
            loop_count++;
        }
        agent.recv();
        agent.send();
    }

    out_pipe.write(END_TOKEN);

    return 0;
}